### PR TITLE
filter inventory at source

### PR DESCRIPTION
**Pull request by bam241**
@scopatz could you tell me if this is acceptable.

this reduce from 60Go++ to be loaded in the memory to 80Mo to load the inventory of a single facility...

if this is right thing to do, **plz do not merge**, I'll rework all my filters to match that, this is way faster and lighter...

### CODE REVIEW

File: cymetric/filters.py [hunk 1]
```diff
@@ -175,20 +175,26 @@ def inventories(evaler, facilities=(), nucs=()):
...
         nucs = format_nucs(nucs)
         rdc_table.append(['NucId', nucs])
 
+    conditionnal_agent = []
+    df = pd.DataFrame()
     if len(facilities) != 0:
         agents = agents[agents['Prototype'].isin(facilities)]
-        rdc_table.append(['AgentId', agents['AgentId'].tolist()])
+        dfs = []
+        for id in agents['AgentId'].tolist():
```
Comment (scopatz) [inline cymetric/filters.py]: You shouldn't name variables `id`, since `id()` is a Python builtin function

File: cymetric/filters.py [hunk 2]
```diff
@@ -175,20 +175,26 @@ def inventories(evaler, facilities=(), nucs=()):
...
+            conditionnal_agent = ('AgentId','==',id)
+            dfs.append(evaler.eval('ExplicitInventory', conds=[conditionnal_agent]))
+            df = pd.concat(dfs)
```
Comment (scopatz) [inline cymetric/filters.py]: This seems like it would take a lot more time to perform. Have you tested that?

**Comment by scopatz** [comment]
Thanks for putting this in @bam241! What were you doing to see those memory reductions? Was that SQLite or HDF5?

File: cymetric/filters.py (hunk 2 above)
Comment (bam241) [inline cymetric/filters.py]: yes I did test it,
the thing is on small simulation, the time different is not noticeable , on big simulation I never reach that point as I was not able to load the full inventory table...

Is there a way to have `OR` condition in this ? I am looping over the conditions just because when I put multiple conditions it is an "AND"...

**Comment by bam241** [comment]
To se those memory reduction I was analysis an cyclus output file (in SQLite) that was 11Go big. and trying to access the timeseries of the inventory of a single archetypes. 
The problem is: I have about 500 agents in my simulation, and are tracking about 200 nuclides, so the inventory table is huge...

File: cymetric/filters.py (hunk 2 above)
Comment (scopatz) [inline cymetric/filters.py]: By design, cyclus backends only AND conditions together.  How big are these files? Out of curiosity...

Comment (bam241) [inline cymetric/filters.py]: 7.7G

Comment (bam241) [inline cymetric/filters.py]: any idea about avoiding the looping ?

**Comment by scopatz** [comment]
Ahh I see... Ok.  We don't really have any tools at all for dealing with large file sizes like this.  The strategy that you have here has the kernel of the right approach (ie to split up the space and the stream operations).

**Comment by scopatz** [comment]
But I think those kinds of splitting / streaming operations should be pushed down to the cyclus backend layer.

**Comment by bam241** [comment]
I am not sure to understand.
Do you mean the looping filter and concatenation part ?

**Comment by scopatz** [comment]
Yes, basically, the cyclus backend interface should provide a way to iterate or stream through a table, in addition to providing the whole table.  While you can achieve some of the benefits of streaming via repeatedly filtering (as you did here), it is 

1. not a general solution because you need to know all of the values to filter over,
2. can have serious performance impacts since on O(N) operation is turned into an O(N^2) [naive] or O(N log(N)) [with indexing]
3. can still have catastrophically large data sets loaded into memory (ie what if one agent has 99% of the inventory objects) 

Besides, filtering is for a fundamentally different operation. Filtering should be used when you don't actually want all of the rows in the result. Streaming is when you don't want the entire result in memory at the same time. Furthermore, you can both stream and filter!

So tl;dr I don't think that the approach used here is robust enough and for large db we probably need some generic interface on the QueryableBackend.
